citest/reporting/simplify_entity_transforms.py

```diff
@@ -0,0 +1,264 @@
+      return label, value
+    if label != 'Path Values' and label != 'Value Justifications':
+      return label, value
+    return label, value[0]
+
+
+class NumericIndexedLabelValueTransformer(EdgeLabelValueTransformer):
+  """Base class for handling results that use '[n]' indexed edge labels.
+
+  This base class determines if there is only one element in the array.
+  If so, it can unpack that one element value to replace the array container.
+  """
+
+  @property
+  def has_exactly_one(self):
+    """True if there is only a [0] label for the array."""
+    return self.__exactly_one
+
+  def __init__(self, key_name, one_label_name, entity, entity_manager):
+    """Constructor.
+
+    Args:
+      key_name: [string] The label name for the edge to the value array.
+      one_label_name: [string] The label name to use for the edge to the value
+         if there is only one value and we replace the array with it.
+    """
+    super(NumericIndexedLabelValueTransformer, self).__init__(
+        entity, entity_manager)
+
+    results = self._find_edge(self.entity, key_name)
+    one = None
+    zero = None
+    if results is not None:
+      entity_id = results.get('_to')
+      results_entity = self.entity_manager.lookup_entity_with_id(entity_id)
+      zero = self._find_edge(results_entity, '[0]')
+      one = self._find_edge(results_entity, '[1]')
+
+    self.__key_name = key_name
+    self.__one_label_name = one_label_name
+    self.__exactly_one = zero is not None and one is None
+
+  def __call__(self, edge):
+    """Implements EdgeLabelValueTransformer.
+
+    If there is only one value and this is the edge to the array containing it,
+    then treat the edge (to the array) as if it were the inner edge
+    (to the value).
+    """
+    label, value = super(
+        NumericIndexedLabelValueTransformer, self).__call__(edge)
+    if not self.__exactly_one or label != self.__key_name:
+      return label, value
+
+    entity_id = edge.get('_to')
+    results_entity = self.entity_manager.lookup_entity_with_id(entity_id)
+    zero = self._find_edge(results_entity, '[0]')
+    value_id = zero.get('_to')
+    value = {'_type': 'EntityReference', '_id': value_id}
+
+    return (self.__one_label_name, value) if value else (None, value)
+
+
+class SequencedResultLabelValueTransformer(
+    NumericIndexedLabelValueTransformer):
+  """Handle edges to "SequencedResult" objects.
+
+  If the result contains exactly one value, replace the array of values with
+  just the single value to get rid of the array indirection.
+  """
+  def __init__(self, entity, entity_manager):
+    """Constructor."""
+    super(SequencedResultLabelValueTransformer, self).__init__(
+        'Results', 'Result', entity, entity_manager)
+
+  def __call__(self, edge):
+    """Implements EdgeLabelValueTransformer."""
+    if self.has_exactly_one:
+      # Drop the count attribute.
+      label = edge.get('label')
+      if label == '#':
```

Comment:
The journal documents some things, but I havent formalized everything yet. I tried not to make any specific for predicate types, though not all predicates use each. Some motivated introducing something, but I tried to tie them to reporting (or data model) concepts as opposed to predicate implementations.